server/src/main/java/io/unitycatalog/server/service/VolumeService.java

```diff
@@ -2,53 +2,175 @@
+import static io.unitycatalog.server.model.SecurableType.SCHEMA;
+import static io.unitycatalog.server.model.SecurableType.VOLUME;
 
 @ExceptionHandler(GlobalExceptionHandler.class)
 public class VolumeService {
   private static final VolumeRepository VOLUME_REPOSITORY = VolumeRepository.getInstance();
+  private static final SchemaRepository SCHEMA_REPOSITORY = SchemaRepository.getInstance();
+  private static final CatalogRepository CATALOG_REPOSITORY = CatalogRepository.getInstance();
 
-  public VolumeService() {}
+  private final UnityCatalogAuthorizer authorizer;
+  private final UnityAccessEvaluator evaluator;
+
+  @SneakyThrows
+  public VolumeService(UnityCatalogAuthorizer authorizer) {
+    this.authorizer = authorizer;
+    evaluator = new UnityAccessEvaluator(authorizer);
+  }
 
   @Post("")
-  public HttpResponse createVolume(CreateVolumeRequestContent createVolumeRequest) {
+  // TODO: for now, we are not supporting CREATE VOLUME or CREATE EXTERNAL VOLUME privileges
+  @AuthorizeExpression("""
+          #authorize(#principal, #metastore, OWNER) ||
+          #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #schema, OWNER, USE_SCHEMA)
+          """)
+  @AuthorizeKey(METASTORE)
+  public HttpResponse createVolume(@AuthorizeKeys({
+                                      @AuthorizeKey(value = SCHEMA, key = "schema_name"),
+                                      @AuthorizeKey(value = CATALOG, key = "catalog_name")
+                                    })
+                                   CreateVolumeRequestContent createVolumeRequest) {
     // Throw error if catalog/schema does not exist
-    return HttpResponse.ofJson(VOLUME_REPOSITORY.createVolume(createVolumeRequest));
+    VolumeInfo volumeInfo = VOLUME_REPOSITORY.createVolume(createVolumeRequest);
+    initializeAuthorizations(volumeInfo);
+    return HttpResponse.ofJson(volumeInfo);
   }
 
   @Get("")
+  @AuthorizeExpression("#defer")
   public HttpResponse listVolumes(
       @Param("catalog_name") String catalogName,
       @Param("schema_name") String schemaName,
       @Param("max_results") Optional<Integer> maxResults,
       @Param("page_token") Optional<String> pageToken,
       @Param("include_browse") Optional<Boolean> includeBrowse) {
-    return HttpResponse.ofJson(
-        VOLUME_REPOSITORY.listVolumes(
-            catalogName, schemaName, maxResults, pageToken, includeBrowse));
+    ListVolumesResponseContent listVolumesResponse = VOLUME_REPOSITORY.listVolumes(
+            catalogName, schemaName, maxResults, pageToken, includeBrowse);
+
+    filterVolumes("""
+            #authorize(#principal, #metastore, OWNER) ||
+            (#authorizeAll(#principal, #schema, OWNER, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG)) ||
+            (#authorize(#principal, #schema, USE_SCHEMA) && #authorizeAny(#principal, #catalog, OWNER, USE_CATALOG) && #authorizeAny(#principal, #volume, OWNER, READ_VOLUME))
```

Comment:
Principals who can get a table: metastore owner, catalog owner, schema owner with USE_CATALOG, and any principal with USE_CATALOG, USE_SCHEMA and any privilege on the table